[Executorch][Quantization][BE] Refactor Choose Qparams

JacobSzwejbka opened the pull request:
Summary: Should hopefully be a little faster. Definitely cleaner to not create an observer inside the op

Test Plan: ci

Differential Revision: D42154677

jerryzh168 commented on torch/ao/quantization/fx/_decomposed.py:
```diff
@@ -200,16 +201,13 @@ def choose_qparams_tensor(
     """
     assert input.dtype == torch.float32, f"Expecting input to have dtype torch.float32, but got dtype: {input.dtype}"
-    assert quant_min < quant_max, f"Expecting quant_min to be smaller than quant_max but received min: {quant_min} max: {quant_max}"
+    assert qmin < qmax, f"Expecting quant_min to be smaller than quant_max but received min: {qmin} max: {qmax}"
```
maybe we can reuse the validation code as well: https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/observer.py#L272

jerryzh168 reviewed (approved):
looks good, thanks!

JacobSzwejbka committed:
[Executorch][Quantization][BE] Refactor Choose Qparams (#92592)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/92592

Should hopefully be a little faster. Definitely cleaner to not create an observer inside the op

Test Plan: ci

Reviewed By: jerryzh168

Differential Revision: D42154677

fbshipit-source-id: 2ab0d6a554e2675b8c5feb7c9c9c4755554fa2b2

JacobSzwejbka commented:
@pytorchbot merge -f

JacobSzwejbka commented:
@pytorchbot --help

JacobSzwejbka commented:
@pytorchbot merge -f the singular failing test failed to random networking issue in the statistics upload stage, the actual test passed.

JacobSzwejbka commented:
@pytorchbot merge -f ' the singular failing test failed to random networking issue in the statistics upload stage, the actual test passed.'

jerryzh168 commented on torch/ao/quantization/utils.py:
```diff
@@ -476,6 +477,116 @@ def _normalize_kwargs(func: Callable, loc: Dict[str, Any]) -> "OrderedDict[str,
+        # setting zero_point to (-1 * min *inv_scale) we get
+        # Xq = Round((Xf - min) * inv_scale)
+        zero_point = -1 * min_val / scale
+    else:
```
should we check qscheme here as well?

atalman commented:
@pytorchbot revert -m "Breaks internal tests" -c ghfirst

atalman commented:
@pytorchbot revert -m "Breaks internal tests" -c ghfirst

atalman commented:
@pytorchbot revert -m "Breaks internal tests" -c ghfirst

malfet commented:
@bigfootjon  @clee2000 not sure why this test were able to merge via GHF, but could not be reverted via GHF. Had something changed with the signals? (As internal PR was abandoned after landing)

malfet commented:
Test that failed to pass internally is this one:
https://github.com/pytorch/pytorch/blob/afe6ea884fa21d424d614efff6bd3a75ec4357af/test/quantization/jit/test_ondevice_quantization.py#L527
```
RuntimeError: isGenericDict() INTERNAL ASSERT FAILED at "fbcode/caffe2/aten/src/ATen/core/ivalue_inl.h":1969, please report a bug to PyTorch. Expected GenericDict but got Int ()
  File "/usr/local/fbcode/platform010/lib/python3.8/unittest/case.py", line 60, in testPartExecutor
    yield
  File "/usr/local/fbcode/platform010/lib/python3.8/unittest/case.py", line 676, in run
    self._callTestMethod(testMethod)
  File "/usr/local/fbcode/platform010/lib/python3.8/unittest/case.py", line 633, in _callTestMethod
    method()
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/buck-out/v2/gen/fbcode/544943cfd42425fb/caffe2/test/quantization/__test_quantization__/test_quantization#link-tree/caffe2/test/quantization/jit/test_ondevice_quantization.py", line 529, in test_device_side_api
    self._check_device_side_api(model)
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/buck-out/v2/gen/fbcode/544943cfd42425fb/caffe2/test/quantization/__test_quantization__/test_quantization#link-tree/caffe2/test/quantization/jit/test_ondevice_quantization.py", line 498, in _check_device_side_api
    self._check_serdes_and_device_side_api_helper(model, True)
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/buck-out/v2/gen/fbcode/544943cfd42425fb/caffe2/test/quantization/__test_quantization__/test_quantization#link-tree/caffe2/test/quantization/jit/test_ondevice_quantization.py", line 435, in _check_serdes_and_device_side_api_helper
    m = _load_for_lite_interpreter(buffer)  # Error here
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/buck-out/v2/gen/fbcode/544943cfd42425fb/caffe2/test/quantization/__test_quantization__/test_quantization#link-tree/torch/jit/mobile/__init__.py", line 50, in _load_for_lite_interpreter
    cpp_module = torch._C._load_for_lite_interpreter_from_buffer(f.read(), map_location)
```

bigfootjon commented:
> @bigfootjon @clee2000 not sure why this test were able to merge via GHF, but could not be reverted via GHF. Had something changed with the signals? (As internal PR was abandoned after landing)

I don't see why this failed to revert. What logic to mergebot use to determine if something was landed via a Diff?
